File: lib/biz/spec-iot/src/main/java/site/addzero/biz/spec/iot/tdengine/TdColumnSpec.java
package site.addzero.biz.spec.iot.tdengine;

import java.util.Locale;
import java.util.Objects;

/**
 * TDengine column specification.
 */
public final class TdColumnSpec {

    public static final String TYPE_TINYINT = "TINYINT";
    public static final String TYPE_INT = "INT";
    public static final String TYPE_FLOAT = "FLOAT";
    public static final String TYPE_DOUBLE = "DOUBLE";
    public static final String TYPE_BOOL = "BOOL";
    public static final String TYPE_NCHAR = "NCHAR";
    public static final String TYPE_VARCHAR = "VARCHAR";
    public static final String TYPE_TIMESTAMP = "TIMESTAMP";
    public static final String NOTE_TAG = "TAG";

    private final String field;
    private final String type;
    private final Integer length;
    private final String note;

    public TdColumnSpec(String field, String type, Integer length, String note) {
        if (field == null || field.trim().isEmpty()) {
            throw new IllegalArgumentException("field must not be blank");
        }
        if (type == null || type.trim().isEmpty()) {
            throw new IllegalArgumentException("type must not be blank");
        }
        this.field = field.trim().toLowerCase(Locale.ROOT);
        this.type = type.trim().toUpperCase(Locale.ROOT);
        this.length = length;
        this.note = note == null ? null : note.trim();
    }

    public String getField() {
        return field;
    }

    public String getType() {
        return type;
    }

    public Integer getLength() {
        return length;
    }

    public String getNote() {
        return note;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof TdColumnSpec)) {
            return false;
        }
        TdColumnSpec that = (TdColumnSpec) other;
        return field.equals(that.field)
                && type.equals(that.type)
                && Objects.equals(length, that.length)
                && Objects.equals(note, that.note);
    }

    @Override
    public int hashCode() {
        return Objects.hash(field, type, length, note);
    }
}

File: lib/biz/spec-iot/src/main/kotlin/site/addzero/biz/spec/iot/tdengine/TdColumnSpec.kt
package site.addzero.biz.spec.iot.tdengine

import site.addzero.biz.spec.iot.requireText
import site.addzero.biz.spec.iot.trimToNull
import java.util.Locale

/**
 * TDengine column specification.
 */
class TdColumnSpec(
    field: String?,
    type: String?,
    val length: Int?,
    note: String?,
) {

    val field: String = requireText(field, "field").lowercase(Locale.ROOT)
    val type: String = requireText(type, "type").uppercase(Locale.ROOT)
    val note: String? = trimToNull(note)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is TdColumnSpec) {
            return false
        }
        return field == other.field &&
            type == other.type &&
            length == other.length &&
            note == other.note
    }

    override fun hashCode(): Int {
        var result = field.hashCode()
        result = 31 * result + type.hashCode()
        result = 31 * result + (length?.hashCode() ?: 0)
        result = 31 * result + (note?.hashCode() ?: 0)
        return result
    }

    companion object {
        @JvmField
        val TYPE_TINYINT: String = "TINYINT"

        @JvmField
        val TYPE_INT: String = "INT"

        @JvmField
        val TYPE_FLOAT: String = "FLOAT"

        @JvmField
        val TYPE_DOUBLE: String = "DOUBLE"

        @JvmField
        val TYPE_BOOL: String = "BOOL"

        @JvmField
        val TYPE_NCHAR: String = "NCHAR"

        @JvmField
        val TYPE_VARCHAR: String = "VARCHAR"

        @JvmField
        val TYPE_TIMESTAMP: String = "TIMESTAMP"

        @JvmField
        val NOTE_TAG: String = "TAG"
    }
}
